Thomah/boothby
# Add authentication

## Status
**READY**

## Migrations
YES
a collection 'user' has been added.

## Description
Added an authentication system in the app.
The passwords are encrypted.
Each API call will be redirected to the auth page, if the user is not authenticated.
To check if a user is authenticated, a token is stored in the server cache, and this token is stored in the session (client-side). Each request will send have this token in the request header.
A function has been written to override the XMLHttpRequest calls, and all the calls have been rewritten using this function.
Schemas of the general behaviour are being drawn. 

## Todos


## Deploy Notes
Notes regarding deployment the contained body of work.  These should note any
db migrations, etc.

## Steps to Test or Reproduce

Outline the steps to test or reproduce the PR here.

You must have a user to be able to auth a first time.
Add a collection user in the mongodb.
In the router.js file, uncomment the line 441, and comment the line 440 : 
you should be able to add a user via localhost:8080/add-user.html.
(We unlock the route /api/user with POST method) 
Once the user is created, put the lines the way they were before, restart the server, and you can log in to the application.

```sh
git pull --prune
git checkout <feature_branch>
```
## Schemas

![general behaviour authentification](https://user-images.githubusercontent.com/8972383/52777481-36fd7380-3044-11e9-91f1-42d72ffcef85.png)

![authentication](https://user-images.githubusercontent.com/8972383/52945443-071adc80-3372-11e9-8078-21a7c35f1281.png)

![api calls 1](https://user-images.githubusercontent.com/8972383/52777560-6613e500-3044-11e9-8072-2d9ed74229c4.png)


![add user](https://user-images.githubusercontent.com/8972383/52777472-3369ec80-3044-11e9-8371-7eaeb28c6198.png)
